src/snowflake/cli/_plugins/nativeapp/entities/application.py

```diff
@@ -634,6 +591,75 @@ def get_objects_owned_by_application(self) -> List[ApplicationOwnedObject]:
             ).fetchall()
             return [{"name": row[1], "type": row[2]} for row in results]
 
+    def _upgrade_app(
+        self,
+        stage_fqn: str,
+        install_method: SameAccountInstallMethod,
+        event_sharing: EventSharingHandler,
+        current_app_row: dict,
+        policy: PolicyBase,
+        interactive: bool,
+    ) -> list[tuple[str]] | None:
+        console = self._workspace_ctx.console
+        self._workspace_ctx.console.step(
+            f"Upgrading existing application object {self.name}."
+        )
+
+        try:
+            return get_snowflake_facade().upgrade_application(
+                name=self.name,
+                current_app_row=current_app_row,
+                install_method=install_method,
+                stage_fqn=stage_fqn,
+                debug_mode=self._entity_model.debug,
+                should_authorize_event_sharing=event_sharing.should_authorize_event_sharing(),
+                role=self.role,
+                warehouse=self.warehouse,
+            )
+        except UpgradeApplicationRestrictionError as err:
+            console.warning(err.message)
+            self.drop_application_before_upgrade(policy=policy, interactive=interactive)
+            return None
+
+    def _create_app(
+        self,
+        stage_fqn: str,
+        install_method: SameAccountInstallMethod,
+        event_sharing: EventSharingHandler,
+        package: ApplicationPackageEntity,
+    ) -> list[tuple[str]]:
+        self._workspace_ctx.console.step(
+            f"Creating new application object {self.name} in account."
+        )
+
+        try:
+            sql_executor = get_sql_executor()
+            if self.role != package.role:
+                with sql_executor.use_role(package.role):
+                    sql_executor.execute_query(
+                        f"grant install, develop on application package {package.name} to role {self.role}"
+                    )
+                    stage_schema = extract_schema(stage_fqn)
+                    sql_executor.execute_query(
+                        f"grant usage on schema {package.name}.{stage_schema} to role {self.role}"
+                    )
+                    sql_executor.execute_query(
+                        f"grant read on stage {stage_fqn} to role {self.role}"
+                    )
+        except ProgrammingError as err:
+            generic_sql_error_handler(err)
```

Comment:
Move this to facade too while we're here :) something like: `grant_create_app` or such. 
Will upgrade our error handling too